### PR TITLE
Install k3s cluster with ansible init

### PR DESCRIPTION
**Pull request by wtripp180901**
NB: currently doesn't support multiple control nodes (needs support for multiple k3s servers, some stuff to do that [here](https://github.com/k3s-io/k3s-ansible/blob/master/roles/k3s_server/tasks/main.yml)

Install k3s cluster on nodes with control plane on control node. Adds a k3s role to run at bootstrap which installs k3s, k9s and helm as well as an ansible-init playbook to start each node's k3s service as an agent or server depending on metadata specified in terraform.

### CODE REVIEW

File: environments/skeleton/{{cookiecutter.environment}}/terraform/compute/variables.tf
Comment (sjpb) [inline environments/skeleton/{{cookiecutter.environment}}/terraform/compute/variables.tf:72]: I don't think we need descriptions in this file, where the vars are just passed through from the outer module

File: ansible/roles/passwords/tasks/main.yml
Comment (sjpb) [inline ansible/roles/passwords/tasks/main.yml:14]: If you see https://developer.hashicorp.com/terraform/language/values/variables#assigning-values-to-root-module-variables, you will see that any vars file ending `.auto.tfvars.json` or `.auto.tfvars` gets automatically loaded.

So rather than regex replacing something in the git-controlled .tfvars file, lets instead *template* a file here like e.g. `ansible-provided.auto.tfvars.json` and template something out using ` ... | to_nice_json` ?

That provides a more general mechanism to set terraform vars from ansible.

Use json so you don't have to template HCL.

File: ansible/roles/passwords/tasks/main.yml
Comment (sjpb) [inline ansible/roles/passwords/tasks/main.yml:16]: Nope. Lets create this in the same place as the other secrets, ansible/roles/passwords/defaults/main.yml

File: environments/common/layouts/everything
Comment (sjpb) [inline environments/common/layouts/everything:88]: Lets change this to `openhpc`. There are cases where the "cluster" (i.e. hosts controlled by this ansible) is wider e.g. extra hosts added to do ... stuff. Bit of a subtle argument TBH.

File: ansible/bootstrap.yml
Comment (sjpb) [inline ansible/bootstrap.yml:262]: This PR needs an image with this included - although I know cross-cloud image handling is hard at present.

File: ansible/bootstrap.yml
Comment (sjpb) [inline ansible/bootstrap.yml:263]:
```suggestion
  become: yes
  tags: k3s
```

File: ansible/cleanup.yml
Comment (sjpb) [inline ansible/cleanup.yml:42]:
```suggestion
- name: Delete ansible-init sentinel file created if ansible-init has run during build
```

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:22]: I think generally accepted practice is to use `loop:` now

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:16]: Can we make this version a rolevar please?

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:36]:
```suggestion
- name: Install k9s
```
Just for consistency

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:26]:
```suggestion
- name: Install helm
```

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:40]: This isn't just "server", right?

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:38]: Hmm. I think this should be separate - so you can add a task to run the k9s role to extras.yml, and in everything groups file default k9s to being control?

File: ansible/roles/k9s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k9s/tasks/main.yml:4]: ideally add owner/group/mode here

File: ansible/roles/k9s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k9s/tasks/main.yml:8]: ideally add owner/group/mode here

Comment (sjpb) [reply ansible/roles/k9s/tasks/main.yml:8]: And later in this file...

File: ansible/roles/k9s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k9s/tasks/main.yml:30]: nice

Comment (sjpb) [reply ansible/roles/k9s/tasks/main.yml:30]: Actually - this is right behaviour but it' be nice to make it idempotent, i.e. add a while, probaly to a block, so that we only do the download etc etc if its not already installed. Maybe using tmp or a tmpfile plays nicer with that? Not sure

**Review by sjpb — CHANGES_REQUESTED** [review]

File: environments/skeleton/{{cookiecutter.environment}}/terraform/nodes.tf
Comment (sjpb) [inline environments/skeleton/{{cookiecutter.environment}}/terraform/nodes.tf:80]: Do we actually need this for the server??

Comment (sjpb) [reply environments/skeleton/{{cookiecutter.environment}}/terraform/nodes.tf:80]: actually, we should be able to ditch the agent type key/value and just run an agent if the server is specified?

**Comment by sjpb** [comment]
@bertiethorpe can you review and merge [this](https://github.com/stackhpc/ansible-slurm-appliance/pull/442) into this PR if you're happy pls? Then update the 1st comment above appropriately.

**Comment by wtripp180901** [comment]
build: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/10957197822

**Comment by wtripp180901** [comment]
new build: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11109726657

**Comment by wtripp180901** [comment]
looks like k3s-agent nodes don't come back up after reboot, don't merge yet

File: environments/skeleton/{{cookiecutter.environment}}/terraform/variables.tf
Comment (sjpb) [inline environments/skeleton/{{cookiecutter.environment}}/terraform/variables.tf:135]: This does need a description here (doesn't in the compute/variables.tf), including that it is set from the ansible auomatically

File: ansible/roles/passwords/tasks/main.yml
Comment (sjpb) [inline ansible/roles/passwords/tasks/main.yml:18]: Given we're making changes here could you just delete this commented out munge-related code pls?

File: ansible/roles/k3s/files/start_k3s.yml
Comment (sjpb) [inline ansible/roles/k3s/files/start_k3s.yml:10]: I think `no_log: true` *prevents* logging stuff? can you check please.

**Review by sjpb — CHANGES_REQUESTED** [review]

File: ansible/roles/k3s/files/start_k3s.yml
Comment (sjpb) [inline ansible/roles/k3s/files/start_k3s.yml:21]:
```suggestion
    - name: Only run loadbalancer on k3s server
    # avoids problems with Ondemand https server
```

File: ansible/roles/k3s/files/start_k3s.yml
Comment (sjpb) [inline ansible/roles/k3s/files/start_k3s.yml:29]:
```suggestion
        - name: Set loadbalancer label on k3s server
```

**Review by sjpb — CHANGES_REQUESTED** [review]

File: ansible/roles/k3s/files/start_k3s.yml
Comment (wtripp180901) [reply ansible/roles/k3s/files/start_k3s.yml:10]: think I forgot to set this back after debugging

File: ansible/roles/passwords/defaults/main.yml
Comment (wtripp180901) [reply ansible/roles/passwords/defaults/main.yml:15]: I thought maybe it shouldn't be there since the ansible doesn't interact with it after this and it made the templating slightly nice but it works as a key too

**Comment by wtripp180901** [comment]
build: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11143143774

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:12]: Not idempotent!

**Review by sjpb — CHANGES_REQUESTED** [review]

**Comment by wtripp180901** [comment]
build: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11177906960

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11238755140

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11276902293

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11288780863

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11520410025

**Comment by wtripp180901** [comment]
don't merge yet, in an azimuth e2e test grafana (cloudalchemy) has a duplicate uid for some reason

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (sjpb) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:415]: Why this change?

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (sjpb) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:419]: OK

File: ansible/roles/k3s/defaults/main.yml
Comment (sjpb) [inline ansible/roles/k3s/defaults/main.yml:5]: Remove this and just use ansible_distribution_major_version in the correct place - this is odd, as it is not a k3s_ prefixed var and I can't see why you would want to override it in its use in ansible/roles/k3s/tasks/main.yml?

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:8]:
```suggestion
- name: Perform air-gapped installation of k3s
```

Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:8]:
```suggestion
- name: Download and air-gapped installation of k3s
  # Using air-gapped install so containers are pre-installed to avoid rate-limiting from registries on cluster startup
```

Is that right?

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:61]: Can we expose the helm version as a k3s_ role var please?

File: environments/.caas/hooks/pre.yml
Comment (sjpb) [inline environments/.caas/hooks/pre.yml:4]:
```suggestion
- name: Generate k3s token
  # NB: Although this generates a new token on each run, the actual token set in metadata is retrieved from a set-once tofu resource, hence only the first value ever generated is relevant.
```

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (sjpb) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:18]:
```suggestion
            input, # makes it a write-once value (set via Ansible)
```

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (sjpb) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:402]: re. using `openstack_compute_instance_v2.control["control"]`, see comments below about changes to caas resources

**Review by sjpb — CHANGES_REQUESTED** [review]
Nearly there, a few comments though.

**Comment by sjpb** [comment]
@wtripp180901 your initial comment on this PR is I think out of date:
> NB: currently requires working DNS to work without requiring reboots. Can currently be worked around by replacing the values of the k3s_server fields in the terraform metadata with openstack_compute_instance_v2.control["control"].access_ip_v4

File: docs/k3s.README.md
Comment (wtripp180901) [reply docs/k3s.README.md:8]: i.e a k3s installation which has previously been configured by ansible-init on boot, as you can't install a new k3s over the top of a previous installation that has already ran. Thinking about it this probably wouldn't be an issue anymore after https://github.com/stackhpc/ansible-slurm-appliance/pull/439 so will remove this

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (wtripp180901) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:402]: I was using the same method to get the control node's IP used as in the changes added in https://github.com/stackhpc/ansible-slurm-appliance/pull/442 to the cookiecutter environment, which require indexed control nodes, otherwise fails with `Because openstack_compute_instance_v2.control does not have "count" or "for_each" set, references to it must not include an index key. Remove the bracketed index to refer to the single instance of this resource.` Could use 
```suggestion
    k3s_server = openstack_compute_instance_v2.control.network[0].fixed_ip_v4
```
instead but probably nice to keep them consistent

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (wtripp180901) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:415]: see below

File: ansible/roles/cluster_infra/templates/outputs.tf.j2
Comment (wtripp180901) [reply ansible/roles/cluster_infra/templates/outputs.tf.j2:27]: see below

**Comment by wtripp180901** [comment]
build https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11662901504

File: ansible/roles/passwords/templates/k3s-token.auto.tfvars.json.j2
Comment (wtripp180901) [reply ansible/roles/passwords/templates/k3s-token.auto.tfvars.json.j2]: making it idempotent means that they will still be mismatched on the initial run because ansible hasn't picked up the templated secrets yet because the role is still running, but they will be updated to be the same on subsequent runs. Added a workaround to fix this

**Comment by wtripp180901** [comment]
^ that merge is just updating my local branch with new password role, current images should be fine to merge if no more changes

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (sjpb) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:415]: Shouldn't make this change

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (sjpb) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:419]: Lets not make changes to caas if we don't actually need to

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (sjpb) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:569]: So this will change, b/c only one control node and potentially no access network defined

File: ansible/roles/cluster_infra/templates/resources.tf.j2
Comment (sjpb) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:402]: - We don't need to keep caas and cookiecutter  consistent if its already different
- caas definitely shouldn't get a set of control nodes (that was a hang-over in cookiecutter from other functionality)
-  I suspect using `openstack_compute_instance_v2.control.network[0].fixed_ip_v4` is fine - it only has 1x network.

File: ansible/roles/passwords/tasks/main.yml
Comment (sjpb) [inline ansible/roles/passwords/tasks/main.yml:10]:
```suggestion
- name: Get templated passwords from target environment
# inventory group/host vars created in a play cannot be accessed in the same play, even after meta: refresh_inventory
```

File: ansible/roles/passwords/tasks/main.yml
Comment (sjpb) [inline ansible/roles/passwords/tasks/main.yml:13]: Lets remove this so it ends up top-level
```suggestion
```

File: ansible/roles/passwords/templates/k3s-token.auto.tfvars.json.j2
Comment (sjpb) [inline ansible/roles/passwords/templates/k3s-token.auto.tfvars.json.j2:2]:
```suggestion
    "k3s_token": "{{ vault_k3s_token }}"
```

now top-level as above

**Review by sjpb — CHANGES_REQUESTED** [review]
Lets minimise changes to the caas / cluster_infra TF templates, then call it done.

File: ansible/roles/k3s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k3s/tasks/main.yml:1]: can we change this to install.yml, just makes it clearer

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11799151926

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/11859638055

File: ansible/extras.yml
Comment (sjpb) [inline ansible/extras.yml:43]: should add a tag here

File: ansible/roles/k9s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k9s/tasks/main.yml:6]:
```suggestion
    register: _k9s_stat_result
```

File: ansible/roles/k9s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k9s/tasks/main.yml:17]:
```suggestion
      when: not _k9s_stat_result.stat.exists
```

File: ansible/roles/k9s/tasks/main.yml
Comment (sjpb) [inline ansible/roles/k9s/tasks/main.yml:44]:
```suggestion
    when: not _k9s_stat_result.stat.exists
```

File: ansible/extras.yml
Comment (sjpb) [inline ansible/extras.yml:42]:
```suggestion
  hosts: k9s
  tags: k9s
```

**Review by sjpb — CHANGES_REQUESTED** [review]
LGTM but I have a couple of minor amends you can just accept *if you are happy*, no need to rebuild

**Review by sjpb — APPROVED** [review]
LGTM

**Pull request merged**
